core/domain/auth_jobs_one_off.py

```diff
@@ -39,6 +39,258 @@
+    without any ERRORs.
+
+    This job is idempotent.
+
+    Pre-conditions:
+        - The models on Oppia reflect the most up-to-date values. Specifically,
+          a user is considered to "exist" IF AND ONLY IF a UserSettingsModel
+          exists.
+        - Exactly 1 FirebaseSeedModel exists.
+        - SeedFirebaseOneOffJob has succeeded AT LEAST ONCE in the history of
+          the database. This includes backups of the database created _after_
+          the job is run.
+        - PopulateFirebaseAccountsOneOffJob has succeeded AT LEAST ONCE in the
+          history of the database, WITHOUT creating any new accounts, AFTER
+          BOTH:
+              1. SeedFirebaseOneOffJob has succeeded.
+              2. Firebase user authentication has been deployed to the server.
+
+    Post-conditions:
+        - There is a 1-to-1-to-1 mapping between UserAuthDetailsModel,
+          UserIdByFirebaseAuthIdModel, and individual Firebase accounts; using
+          firebase_auth_id as the key. The sole exception is:
+              UserAuthDetailsModel(gae_id=feconf.SYSTEM_COMMITTER_ID)
+        - UserAuthDetailsModel.deleted,
+          UserIdByFirebaseAuthIdModel.deleted, and
+          firebase_admin.auth.ExportedUserRecord.disabled each have the same
+          respective boolean value.
+      """
+
+    SYSTEM_COMMITTER_ACK = 'INFO: SYSTEM_COMMITTER_ID skipped'
+
+    EMPTY_AUTH_ID_KEY = '<EMPTY_AUTH_ID_KEY>'
+    ASSOC_BY_AUTH_ID_KEY = 'assoc_by_auth_id'
+    ASSOC_BY_USER_ID_KEY = 'assoc_by_user_id'
+    FIREBASE_ACCOUNT_DISABLED_KEY = 'firebase_account_disabled'
+
+    ERROR_ASSOC_INCONSISTENCIES = (
+        'ERROR: Found inconsistency in models and/or Firebase account')
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [auth_models.UserIdByFirebaseAuthIdModel,
+                auth_models.UserAuthDetailsModel,
+                auth_models.FirebaseSeedModel]
+
+    @staticmethod
+    def map(item):
+        # The map() function must be static, so we manually create a "cls"
+        # variable instead of changing the function into a classmethod.
+        cls = SyncFirebaseAccountsOneOffJob
+
+        if isinstance(item, auth_models.UserIdByFirebaseAuthIdModel):
+            auth_id, assoc_info = (
+                item.id,
+                (cls.ASSOC_BY_AUTH_ID_KEY, (item.user_id, item.deleted)))
+            yield (auth_id, assoc_info)
+            return
+
+        if isinstance(item, auth_models.UserAuthDetailsModel):
+            if item.gae_id == feconf.SYSTEM_COMMITTER_ID:
+                yield (cls.SYSTEM_COMMITTER_ACK, item.id)
+                return
+            auth_id, assoc_info = (
+                item.firebase_auth_id or cls.EMPTY_AUTH_ID_KEY,
+                (cls.ASSOC_BY_USER_ID_KEY, (item.id, item.deleted)))
+            yield (auth_id, assoc_info)
+            return
+
+        for user in firebase_admin.auth.list_users().iterate_all():
```

Comment:
Maybe add a comment that this corresponds to FirebaseSeedModel -- took me a little while to grok that.

Reply:
Done.